Added PlaybackPolicy type

ancorso opened the pull request:
The PlaybackPolicy type allows for the replay of a fixed sequence of actions on an MDP or POMDP. Users must also specify a backup policy that is used if the episode is still going after the fixed sequence of actions is applied.

ancorso committed:
Added new PlaybackPolicy type that allows for the replay of a fixed sequence of actions on an MDP or POMDP.

MaximeBouton commented on src/playback.jl:
```diff
@@ -0,0 +1,45 @@
+
+"""
+    RandomPolicy{RNG<:AbstractRNG, P<:Union{POMDP,MDP}, U<:Updater}
```
typo, I think it should be PlaybackPolicy

ancorso replied:
Yes, that is a typo, thanks!

MaximeBouton commented on src/playback.jl:
```diff
@@ -0,0 +1,45 @@
+"""
+mutable struct PlaybackPolicy{A} <: Policy
+    actions::Vector{A}
+    backup_policy::Policy
```
the policy type should be part of the parameters to make it a concrete type: 
```julia
struct PlaybackPolicy{A, P<:Policy} <: Policy 
    actions::Vector{A}
    backup_policy::P
    logpdfs::Vector{Float64} 
    i::Int64
end
```

MaximeBouton replied:
Even better would be to use `AbstractArray` instead of `Vector` for `actions` and `logpdfs`

ancorso replied:
Agreed, thanks for pointing that out!

MaximeBouton commented on src/playback.jl:
```diff
@@ -0,0 +1,45 @@
+end
+
+# Constructor for the PlaybackPolicy
+PlaybackPolicy(actions::Vector{A}, backup_policy::Policy; logpdfs::Vector{Float64} = Float64[]) where {A} = PlaybackPolicy(actions, backup_policy, logpdfs, 1)
```
`A` is not used in the function so it might be removed.

MaximeBouton replied:
```
# Constructor for the PlaybackPolicy
PlaybackPolicy(actions::AbstractArray, backup_policy::Policy; logpdfs::AbstractArray{<:Real} = Float64[]) = PlaybackPolicy(actions, backup_policy, logpdfs, 1)
```

ancorso replied:
Ok, removed A from the type

MaximeBouton commented:
The `PlaybackPolicy` is using a vector to represent the list of action, and stores the index another possibility is to use an iterator, this might allow to use more general types than vectors, and vectors would still work. Up to you to try it out, it is ok if we merge this PR with the vector approach.

ancorso committed:
Relaxed types on actions and logpdfs, fixed typo in PlaybackPolicy description, Removed type parameter A from constructor

ancorso commented:
Addressed all comments in the review. Thanks for looking it over. I agree using an iterator would probably be better but I'm ok leaving that to future work for now.

MaximeBouton commented on src/playback.jl:
```diff
@@ -0,0 +1,45 @@
+mutable struct PlaybackPolicy{A, P<:Policy, V<:Real} <: Policy
+    actions::AbstractArray{A}
+    backup_policy::P
+    logpdfs::AbstractArray{V}
```
sorry, my previous comment was a bit unclear, if you use an abstract type it should be parameterized: 

```julia 
mutable struct PlaybackPolicy{A<:AbstractArray, P<:Policy, V<:AbstractArray{<:Real}} <: Policy
    actions::A
    backup_policy::P
    logpdfs::V
    i::Int64
end
```

ancorso replied:
Thanks for pointing this out. I was unclear about the distinction between concrete and abstract types.

ancorso committed:
Defined concrete types for the struct members
